tempesta_fw/tls.c

```diff
@@ -22,319 +22,548 @@
 #include "cfg.h"
 #include "connection.h"
 #include "client.h"
+#include "msg.h"
+#include "procfs.h"
 #include "tls.h"
 
-#define tls_dbg(c, fmt, ...)					\
-	TFW_DBG("TLS(%p,%p):%s " fmt "\n", c, c->sk, __func__, ## __VA_ARGS__)
-#define tls_err(c, fmt, ...)					\
-	TFW_ERR("TLS(%p,%p):%s " fmt "\n", c, c->sk, __func__, ## __VA_ARGS__)
+static struct {
+	ttls_config	cfg;
+	ttls_x509_crt	crt;
+	ttls_pk_context	key;
+	unsigned long	crt_pg_addr;
+	unsigned int	crt_pg_order;
+} tfw_tls;
 
-#define MAX_TLS_PAYLOAD		1460
-
-typedef struct {
-	mbedtls_ssl_config	cfg;
-	mbedtls_x509_crt	crt;
-	mbedtls_pk_context	key;
-} TfwTls;
-
-static TfwTls tfw_tls;
-
-static inline int
-ttls_ssl_handshake(TfwTlsContext *tls)
+/**
+ * Chop skb list with begin at @skb by TLS extra data at the begin and end of
+ * the the list after decryption and write the right pointer at the first skb
+ * and offset to @data for upper layers processing.
+ */
+static int
+tfw_tls_chop_skb_rec(TlsCtx *tls, struct sk_buff *skb, unsigned int off,
+		     TfwFsmData *data)
 {
-	int r;
+	size_t prolog = ttls_payload_off(&tls->xfrm);
+
+eaten_skb:
+	if (likely(skb->len > prolog)) {
+		off += prolog;
```

Comment:
It seems that case with `skb->len > prolog` and `skb->len == prolog + off` may appear here, i.e. skb's space is exhausted but we will continue to deal with it.

Reply:
Now we chop the possible TCP seqnos overlap at the begin of `tfw_tls_msg_process()` because we can not link such overlapping skb at the middle of the skb chain for HTTP processing (there is no way to specify offsets for each skb in the list). So `prolog` went away and `off` is TLS header and IV.